### ISSUE TITLE
Overfitting and increasing class loss

### ISSUE THREAD

**Issue by ArpanGyawali** [MIC-DKFZ/nnDetection#290]
I am working on LVO detection using 3d MIP images. Half of the image have no annotations as they are negative case, but are important for training to learn negative features. And the other half have 1 and sometimes 2 LVO segmentation associated with each image for positive case. The segmentations from 2 to 3 slices.
As i am training the training loss goes negative after 18-19 epochs(Total 25 epochs) and the validation loss starts increasing after very few epochs. I think this is due to overfitting.
I have around 220 training data with (80:20) train val split. That means each fold contains only 44 data. Is this because of less training data. 
Also, when looking at the validation class loss, its increasing rapidly after very few epochs, while the regression loss starts at negative value and gradually decrease. Why is this?
![image](https://github.com/user-attachments/assets/154bd102-9fe4-4388-aad8-e84a6b09b49b)
![image](https://github.com/user-attachments/assets/32987bcf-a91f-45be-aed5-122eddbde704)

When i use some more augmentation, then the validation loss increase by a bit (i.e. from 0.7 to 0.5), but still its not good and overfitting.
![image](https://github.com/user-attachments/assets/3743cd7f-bca0-4e73-b565-6c60e4c65d7f)

Doesnt nndet use focal loss as class loss?  
Can you give some insights to how i can increase acccuracy of my model. If more information is required i can provide.

**Comment by andres2631996** [comment]
By default, nnDetection utilizes BCE as class loss. You could try to use focal loss adding the following snippet to /nndet/ptmodule/retinaunet/v001.py

```
from nndet.arch.heads.classifier import FocalClassifier

@MODULE_REGISTRY.register
class RetinaUNetV001Focal(RetinaUNetModule):
    base_conv_cls = ConvInstanceRelu
    head_conv_cls = ConvGroupRelu

    head_cls = DetectionHeadHNMNative
    head_classifier_cls = FocalClassifier
    head_regressor_cls = GIoURegressor
    matcher_cls = ATSSMatcher
    segmenter_cls = DiCESegmenterFgBg
```

And then run:

`nndet_train TASK -o exp.fold=0 module=RetinaUNetV001Focal`


It may also be that your objects are very small. If the objects you want to detect are just like 3x3x3 in images with hundreds of voxels in each dimension, as you claim, it may be that the imbalance is just very strong. Can you check the size of any of your boxes in _det_data/TASK/preprocessed/PLAN/imagesTr/CASEID_boxes.pkl_?

If you are working with LVOs, you could define larger ROIs in the segmentations in _raw_splitted/labelsTs_ to alleviate the imbalance. LVOs are just regions of interest, unlike tumors or other lesion types with a defined boundary, and that allows for flexibility in its object definition

**Comment by ArpanGyawali** [comment]
Thank you so much. @andres2631996 
I tried the focal loss as you said but the ap@0.1 as well as dice score doesn't move more than 0.55 with average of 0.5.
The loss goes negative just after 3-4 epoch and overfitting  has reduced by a little but. But the results are still not good enough.
 
The config is something like this,
```
# @package **global**
defaults:
  - augmentation: base_more
module: RetinaUNetV001
predictor: BoxPredictorSelective
plan: D3V001_3d # plan used for training
planner: D3V001 # planner used for preprocessing
  
augment_cfg:
  augmentation: ${augmentation}
  num_train_batches_per_epoch: ${trainer_cfg.num_train_batches_per_epoch}
  num_val_batches_per_epoch: ${trainer_cfg.num_val_batches_per_epoch}
  dataloader: "DataLoader{}DOffset"
  # oversample_foreground_percent: 0.5 # ratio of fg and bg in batches
  oversample_foreground_percent: 0.75 # increased
  dataloader_kwargs: {}
  num_threads: ${oc.env:det_num_threads, "12"}
  num_cached_per_thread: 2
  multiprocessing: True # only deactivate this if debugging
  # Additional overwrites
  # patch_size; Default: plan
  # batch_size; Default: plan
  # splits; Default: splits_final
trainer_cfg:
  gpus: 2 # number of gpus
  accelerator: # distributed backend
  precision: 16 # mixed precision
  amp_backend: native # mixed precision backend
  amp_level: O1 # when mixed precision backend is APEX use O1
  # Per default training is deterministic, non-deterministic allows
  # cudnn.benchmark which can give up to 20% performance. Set this to false
  # to perform non-deterministic training
  deterministic: False
  benchmark: False  
  monitor_key: "mAP_IoU_0.10_0.50_0.05_MaxDet_100" # used to determine the best model
  monitor_mode: "max" # metric operation mode "min" or "max"
  max_num_epochs: 15 # max number of epochs
  num_train_batches_per_epoch: 400 # number of train batches per epoch
  num_val_batches_per_epoch: 100 # number of val batches per epoch
  initial_lr: 0.01 # initial learning rate to start with (Preiously->0.1)
  sgd_momentum: 0.9 # momentum term
  sgd_nesterov: True # nesterov momentum
  weight_decay: 5.e-4 # weight decay for optimizer(Previous -> 3.e-5)
  warm_iterations: 1500 # number of iterations with warmup
  warm_lr: 1.e-6 # learning rate to start warmup from
  poly_gamma: 0.9
  swa_epochs: 5 # number of epochs to run swa with cyclic learning rate
  # sweep_ckpt: Select checkpoint identifier for sweeping. Default "last".
model_cfg:
  encoder_kwargs: {} # keyword arguments passed to encoder
  decoder_kwargs: # keyword arguments passed to decoder
    min_out_channels: 16         #from 8
    upsampling_mode: "transpose"
    num_lateral: 2               #from 1
    norm_lateral: True           #from false
    activation_lateral: True     #from false
    num_out: 1
    norm_out: True
    activation_out: True
  head_kwargs: {} # keyword arguments to passed to head
  head_classifier_kwargs: # keyword arguments passed to classifier in head
    num_convs: 2
    norm_channels_per_group: 16
    norm_affine: True
    reduction: "mean"
    # loss_weight: 1.
    # prior_prob: 0.01
    loss_weight: 2.
    prior_prob: 0.1
  head_regressor_kwargs: # keyword arguments passed to regressor in head
    num_convs: 2
    norm_channels_per_group: 16
    norm_affine: True
    reduction: "sum"
    loss_weight: 1.5
    learn_scale: True
  head_sampler_kwargs: # keyword arguments passed to sampler
    batch_size_per_image: 8 # number of anchors sampled per image (prev-> 32)
    positive_fraction: 0.6 # defines ratio between positive and negative anchors (Prev -> 0.33)
    # hard negatives are sampled from a pool of size:
    # batch_size_per_image * (1 - positive_fraction) * pool_size
    # pool_size: 20
    pool_size: 8
    min_neg: 1 # minimum number of negative anchors sampled per image
  segmenter_kwargs:
    dice_kwargs:
      batch_dice: True
  matcher_kwargs: # keyword arguments passed to matcher
    num_candidates: 4
    center_in_gt: False
  plan_arch_overwrites: {} # overwrite arguments of architecture
  plan_anchors_overwrites: {} # overwrite arguments of anchors
```

**Comment by andres2631996** [comment]
Hello, everything looks good to me, I do not see any clear red flag in your hyperparameters. I think the overfitting problem may happen due to the strong imbalance in the size of your objects, compared to the overall size of the image. How much is your AP at IoU=0.5 again? The main metric optimized is the mAP at IoU thresholds from 0.10 to 0.50 in 0.05 steps _(mAP_IoU_0.10_0.50_0.05_MaxDet_100)_, so I would focus mostly on this metric to see if your model is really learning something, rather than on the Dice score. Could you let me know your value for _mAP_IoU_0.10_0.50_0.05_MaxDet_100_ again? Is this value after running all cross-validation folds or only for one of the cross-validation folds (before consolidation)?

Could you check in some of the cases where you have LVOs the size of the boxes that are prepared for training the model? You can do this in _det_data/TASK/preprocessed/D3V001_3d/imagesTr/CASEID_boxes.pkl_, where TASK is your task folder with the task ID, and CASEID is some case ID of your own that has LVOs. You can let me know the size of your boxes, because I suspect that maybe they are too small. I work myself with LVO detection, and the models tend to easily overfit. You could also try the _insane_ augmentation scheme, but I do not think that will be a game changer.

**Comment by ArpanGyawali** [comment]
Thank you @andres2631996 ,
I am right now running last fold training and the mAP@0.1:0.5:0.05 is 0.723 at 4th epoch, but in other folds the MAP decreased after few epochs(i.e. 6-8 epochs).

Where can I find the average best MAP@0.1-0.5 after the consolidation?

One of the image siize is array([ 18, 512, 512]))
Where the boxes points are: array([[  5, 195,   8, 212, 313, 335]]
The boxes generally spans or 2-3 slices along the z axis(i.e. along 17-18 slices) and about 20 pixels along x and y axis(along 512)

**Comment by andres2631996** [comment]
The values for mAP after each epoch look ok for me for the task of LVO detection. During training, the mAP is computed in different patches of the validation set without any postprocessing, while when running _nndet_eval_, the results from all the patches from all validation cases are combined together and duplicates removed via non-maximum suppresion and matched to the ground-truth boxes, so the mAP values during the training logs and after running _nndet_eval_ are not the same. What you need to ultimately follow are the values in the _val_results_ folder after running _nndet_eval_

The _mAP_IoU_0.10_0.50_0.05_MaxDet_100_ values in _val_results_ look reasonable to me. In my LVO experiments I tend to get values around 0.6, but on larger datasets than yours.

To run the consolidation, after you have the results for all the folds and you have done _nndet_sweep_ for each fold, run this:
`
nndet_consolidate TASK_ID MODEL --sweep_boxes`

Now that you show me the relative size of your boxes with respect to the image size, I finally get why your boxes in Z only span 3 slices. I thought your images looked more isotropic in Z. Then it should not be such a big problem that your boxes only span 3 slices

Augmentations are completed with _batchgenerators_, and in there the format X, Y, Z is followed for the data, so the _mirror_axes_ hyperparameter makes sense to be [2].  _batchgenerators_ only has 0.5 probability for mirroring, which is the value with the highest variability, so to have a similar amount of flipped and non-flipped patches. Increasing the probability would not be beneficial in my opinion, as you would just end up with more flipped data and less original data, but the variability shown to the model would decrease

**Comment by ArpanGyawali** [comment]
Thank you so much @andres2631996 for the explanation.
I am getting the consistent MAP@0.1-0.5 of 0.7 in fold 4.

I am running training on each fold with the --sweep option already.
You said to look at the val_results folder after running nndet_eval, when should i run this? is it after consolidating with the fold as -1? or before consolidating on each fold(0-4).

**Comment by andres2631996** [comment]
Oh yeah, you need to do:
`nndet_eval TASK MODEL -1` to get the evaluation results in the consolidated folder under the _val_results_ folder

The option _--test_ is for an external test set with images in _det_data/TASK/raw_splitted/imagesTs_ and labels in _det_data/TASK/raw_splitted/labelsTs_, so your reminaing 50 cases should be in these folders

**Comment by ArpanGyawali** [comment]
Thanks @andres2631996 ,
Sorry about the confusion, I actually spitted the images excluding the 50 test cases i mentioned earlier into 80-20 ratio. Where 80% went to imagesTr and 20% to imagesTs. Arent images from imagesTs cases taken as validation set for each epoch or they are not used while training?
The 50 images i told you about are different from the images from imagesTr and imagesTs.

If the imagesTs arent used in training and validation, Then I should split them into 80:20 and use all the images for training and the other 50 cases can be used for external testing put manually under imagesTs folder.

**Comment by andres2631996** [comment]
_imagesTs_ are not used neither in training nor in validation, the images in _imagesTr_ are used for both training and validation in cross-validation. _imagesTs_ and _labelsTs_ are folders for hold out test sets that are not observed during model design. This is also the same case with nnUNet, a self-configuring framework from our same division

If you want to run an external hold out test set with those 50 images that you left apart from model design, I would recommend you to leave them in _imagesTs_ (and if you have test labels, leave the test labels in _labelsTs_). Then you can run all the commands from my previous comment

No worries, there are no silly questions :)

**Comment by ArpanGyawali** [comment]
I am clear now. Thank you so much @andres2631996 . This increase my dataset by 54 more images and hope to see some increment in the result.
I will close the issue for now and if I have more ques regarding this, I will reopen it.

**Comment by ArpanGyawali** [comment]
Hi @andres2631996 
Sorry for disturbing you again,
I had one confusion for a long,
What does the confusion matrix under val analysis mean,
![image](https://github.com/user-attachments/assets/06cabd9f-2167-4039-a8ad-fc60f44a01f6)

![image](https://github.com/user-attachments/assets/763fe772-4763-4a42-a7a9-df9851d5611c)

These looked confusing to me.

**Comment by andres2631996** [comment]
The confusion matrix just compares the predictions done by the model vs the ground truth boxes. The predicted boxes are accompanied by a predicted score that indicates how confident the prediction is. Since a match between a predicted box and a ground truth box can be considered at different IoU thresholds and the predicted boxes can be also thresholded based on their respective prediction score, one can create different confusion matrices based on the IoU thresholds and the scores from the boxes.

I d recommend you to mostly look at the confusion matrix generated at IoU threshold 0.1 and prediction box score threshold of 0.5

**Comment by andres2631996** [comment]
There are no true negatives, because detection models do not predict any negative boxes. It makes sense to have many false positives in such an imbalanced task as LVO detection. Probably some of those false positives are repeated predictions

**Comment by ArpanGyawali** [comment]
What do you mean by repeated predictions? Does that mean overlapped predictions? Is this done before nms thresholding?

**Comment by andres2631996** [comment]
Maybe there could be still missed repetitions that may not be handled by NMS. Maybe it is also good to check the nnDetection paper if you have very theoretical questions

**Comment by ArpanGyawali** [comment]
Sure thing, Thank you so much again.

**Comment by ArpanGyawali** [comment]
@andres2631996 Hi,
I have some questions after conducting evaluations and prediction. The sensitivity of the model at 0.4 score threshold is only 58% and that is very low. Still there is overfitting, how can i decrease this overfitting. The training MAP@0.1-0.5 is above 0.7 but on the test set, it goes down to 0.5. Do increasing the amound of data solve this issue. Or is the model very complicated so that it worked good while training but couldnt generalize properly, or is there some other ways to improve my model?
![image](https://github.com/user-attachments/assets/4eb61608-ce38-40d5-8b2a-011a9dcefdb1)
This is the dataset and prediction information. The distal LVO cases are on the lo side and the prediction on the distal region is also low. What would be a good amount of data for this model to work satisfactorily, with atleast 0.8 sensitivity on the test set.

**Comment by andres2631996** [comment]
Hello, are the testing images from another scanner? If so, it is expected to see some decrease in the performance due to domain shifts. Distal and middle cases are usually less frequent to see and rather subtle, that is why the model performs worse with them. I see that there is quite some prevalence shift: the proportion of middle cases in your training set is much higher than in the test set, that could induce some bias. I would say in general that it is not so bad to observe that value for sensitivity, since it is quite a difficult problem. You can try different score thresholds to see where you get the best balance between sensitivity and specificity.


Any kind of data increase would be beneficial for your problem. I cannot come with a specific number right now

**Comment by ArpanGyawali** [comment]
Thank you @andres2631996 
I am asking for 200 more data. This may help
The test data are not from different scanner, they are all mixed, but the test data are totally unseen to the model. 
Also, Other than increasing the number of data, is there any way to deal with this problem? Like any dropout or other regularizatin or any other change in config. I want to keep the data increment my last choice.
This project has took a long for me to finish, and I would love to get help in improving it.

**Comment by andres2631996** [comment]
Hello, thanks for letting me know that there are no domain shifts caused by scanner/hospital/etc. Are all your original segmentations of the LVOs approximately the same size? If so, you could try to modify the anchors in your pkl plan file to only include objects with a similar size. 

If you open the _plan.pkl_ file, could you let me know what is the value for the field _anchors_? You could reduce the widths, heights, and depths to two sizes, like (not the same values, but in the same style):

```
"anchors": {
        "width": [
            [
                8.0,
                15.0
            ],
            [
                16.0,
                30.0
            ],
            [
                32.0,
                60.0
            ],
            [
                64.0,
                120.0
            ]
        ],
        "height": [
            [
                9.0,
                14.0
            ],
            [
                18.0,
                28.0
            ],
            [
                36.0,
                56.0
            ],
            [
                72.0,
                112.0
            ]
        ],
        "depth": [
            [
                9.0,
                14.0
            ],
            [
                18.0,
                28.0
            ],
            [
                36.0,
                56.0
            ],
            [
                72.0,
                112.0
            ]
        ],
        "stride": 1
    },
```

You could also just train to the approximate number of epochs after you see the overfitting starts imposing, by modifying the field _max_num_epochs_ in _nnDetection/tree/main/nndet/conf/train
/v001.yaml_

It is also important that you do not restrain yourself to calculate sensitivity and specificity at a threshold score of 0.4, but that you also see if there are other scores where the relation between sensitivity and specificity would be better than at 0.4

**Comment by ArpanGyawali** [comment]
Hi @andres2631996 
my anchor
'anchors': { 'depth': [(10.0, 12.0, 19.0), (20.0, 24.0, 38.0)],
               'height': [(9.0, 7.0, 13.0), (18.0, 14.0, 26.0)],
               'stride': 1,
               'width': [(4.0, 87.0, 22.0), (4.0, 87.0, 22.0)]}
              
These are the anchor size of my plan.pkl and as mentioned above one of the image siize is array([ 18, 512, 512]))
Where the boxes points are: array([[ 5, 195, 8, 212, 313, 335]]. ALl other image size is same and the boxes sizes are also somehow similar in dimention. i.e.(2-3, 15-20, 20-30) 

So, according to this what numbers would be good for my anchor size as you mentioned? 
After changing this, lowering the number of epochs and using some dropouts, i will retrain the model.

**Comment by andres2631996** [comment]
Maybe removing the last element of the anchor dicrionary in each dimension could help? 

anchors': { 'depth': [(10.0, 12.0), (20.0, 24.0)],
'height': [(9.0, 7.0), (18.0, 14.0)],
'stride': 1,
'width': [(4.0, 87.0), (4.0, 87.0)]}

Not sure if the configuration d work but let s see

You would need to create a new module in nndet.arch.conv to include the dropout, at the same level as BaseConvNormAct. And then modify the function RetinaUNetV001 in nndet.ptmodule.retinaunet. Unfortunately I am out of office and I cannot dive into a lot of details here at the moment.

**Comment by ArpanGyawali** [comment]
Thanks @andres2631996 
No worries about the dropout.
But if you can, can you explain me how the anchor size represenation works?
I mean hat the numbers refers to? The depth and height has the scale of 2 in the 2nd set but not for width. Also why width ranges from 4 to 87?
I couldn't get it.

**Comment by andres2631996** [comment]
The anchor sizes are set bases on your object sizes. Given the high anisotropy of your objects, It is expected that one of the dimensions will be much larger than the other two

**Comment by ArpanGyawali** [comment]
hi @andres2631996 , One more question.
The validation scores are much better than the test scores and predictions. Both val and test data are from similar scans and have same distribution, but why is the model with 0.58 MAP score for validation set while 0.47 on the test set.

Also, the prediction scores provided on the prediction result are calculated on what IOU?

**Comment by ArpanGyawali** [comment]
@andres2631996  Is the IOU used is the one that is ensemble best IOU at sweep inside consolidate folder?
But that says 0.4 IOU and best score of 0.0. I dont understand how this is best as the score of 0.0 gives a lot of false positives.

**Comment by andres2631996** [comment]
You have a larger share of distal cases in the test set than in validation, the difficulty in processing those cases could produce a shift in the scores caused by a prevalence shift

The prediction scores are a probability provided by the detection Network, trained to minimize box overlap at 0.10 IoU

I dont understand your last question. Could you rephrase It?

**Comment by ArpanGyawali** [comment]
Sure @andres2631996 
At the end of consolidation, the below log was printed
```Determined {'model_iou': 0.5, 'model_nms_fn': <function batched_weighted_nms_model at 0x1519d94a0550>, 'model_score_thresh': 0.5, 'model_topk': 1000, 'model_detections_per_image': 100, 'ensemble_iou': 0.4, 'ensemble_nms_fn': <function batched_wbc_ensemble at 0x1519d945fdc0>, 'ensemble_topk': 1000, 'remove_small_boxes': 3.0, 'ensemble_score_thresh': 0.0} with best sweeping score 0.5435654948433765 mAP_IoU_0.10_0.50_0.05_MaxDet_100```
So does this means that the predict command uses IOU 0.4 as default IOU as the best_ensemble IOU is 0.4.

The process from training each folds to consolidation. How the final best model is taken from all those folds to run prediction on the new data. I didnt quite understood this process. I went through the paper but still confused, Because, if the best model is chosen then why is the val_MAP score of the consolidate folder less then that of fold 3 and 4(in my case)?
It would be great if you explain this to me.

**Comment by andres2631996** [comment]
Hello, I think we have to separate two different items:
- The overlap between each predicted box and the matched ground-truth box. Here, an IoU value of 0.10 is used to threshold predictions in the 3D space. _prediction vs ground truth_
- The overlap between predicted boxes by different models of an ensemble. _prediction model X vs prediction model Y_. What you are showing me indicates that when ensembling predictions, those matched predictions that present an IoU < 0.4 are discarded. At the same time, _ensemble_score_thresh_ is a _minimum_ threshold and not _maximum_, meaning to consider in the ensemble those boxes with a probability score > 0.0

In summary, when ensembling boxes that overlap with IoU of 0.4 and with a minimum probability score of 0.0, a maximum _mAp_IoU_0.10_0.50_0.05_MaxDet_100_ in the validation set is obtained, after removing duplicates with NMS

The consolidation process takes each model in the cross-validation setup, runs predictions in the corresponding validation set from each fold, and includes 8 mirroring transforms as Test-Time Augmentation for each case in the validation set of each fold. All the predictions for the same case are ensembled, and different ensemble IoU values and minimum probability scores are tried, taking the combination of hyperparameters maximizing _mAp_IoU_0.10_0.50_0.05_MaxDet_100_ for all those predictions, compared to the ground-truth boxes

The val_MAP score of the consolidated folder cannot be compared to the val_MAP scores of folds 3 and 4, because the val_MAP score of the consolidated folder is computed on the predictions of ALL the validation sets from ALL folds (the different predictions from all folds are _consolidated_ as indicated in the paragraph below). The val_MAP scores of folds 3 and 4 are computed ONLY in the validation sets of folds 3 and 4, respectively

**Comment by ArpanGyawali** [comment]
So the default IOU used for the prediction is 0.1.
Thank you so much @andres2631996 for explaining this to me in easy words. Now i understand the difference. So the only way to improve my model by decreasing over fitting is to get more data and to balance out the dataset with each type of LVO, right? Other approach didn't improve the test result.

**Comment by andres2631996** [comment]
Yes, I think those ideas could work. Unfortunately, I have worked myself with occlusion detection and It s quite a difficult problem, compared to lung nodule detection or fracture detection

**Comment by ArpanGyawali** [comment]
Yes, I worked on 2d Dataset for LVO detection previously, by creating 2d MIPS. I got the sensitivity of around 80-85. But i wanted to try in 3D thinking this might bring the sensitivity to around 85-90. I dont think thats possible but lets see where it takes.
